File: py/tests/test_scalers.py
"""Tests for Python StandardScaler and MinMaxScaler."""

import numpy as np
import pytest

from wlearn.scalers import StandardScaler, MinMaxScaler
from wlearn.bundle import decode_bundle
from wlearn.registry import load as registry_load
from wlearn.errors import NotFittedError, DisposedError, ValidationError


class TestStandardScaler:
    def test_fit_transform_zero_mean_unit_var(self):
        X = np.array([[1, 10], [2, 20], [3, 30], [4, 40]], dtype=np.float64)
        scaler = StandardScaler()
        result = scaler.fit_transform(X)
        assert result.shape == X.shape
        # Column means should be ~0
        assert np.allclose(result.mean(axis=0), 0, atol=1e-10)
        # Column stds should be ~1 (sample std, ddof=1)
        assert np.allclose(result.std(axis=0, ddof=1), 1, atol=1e-10)

    def test_fit_then_transform(self):
        X = np.array([[1, 2], [3, 4], [5, 6]], dtype=np.float64)
        scaler = StandardScaler()
        scaler.fit(X)
        result = scaler.transform(X)
        assert result.shape == X.shape
        assert np.allclose(result.mean(axis=0), 0, atol=1e-10)

    def test_save_load_roundtrip(self):
        X = np.array([[1, 10], [2, 20], [3, 30]], dtype=np.float64)
        scaler = StandardScaler()
        scaler.fit(X)
        result_orig = scaler.transform(X)

        bundle = scaler.save()
        manifest, toc, blobs = decode_bundle(bundle)
        loaded = StandardScaler._from_bundle(manifest, toc, blobs)
        result_loaded = loaded.transform(X)
        assert np.allclose(result_orig, result_loaded)

    def test_registry_load(self):
        X = np.array([[1, 2], [3, 4]], dtype=np.float64)
        scaler = StandardScaler()
        scaler.fit(X)
        bundle = scaler.save()

        loaded = registry_load(bundle)
        assert isinstance(loaded, StandardScaler)
        assert loaded.is_fitted

    def test_constant_column(self):
        X = np.array([[5, 1], [5, 2], [5, 3]], dtype=np.float64)
        scaler = StandardScaler()
        result = scaler.fit_transform(X)
        # Constant column should produce all zeros
        assert np.all(result[:, 0] == 0)

    def test_not_fitted_error(self):
        scaler = StandardScaler()
        with pytest.raises(NotFittedError):
            scaler.transform(np.zeros((1, 2)))

    def test_disposed_error(self):
        X = np.array([[1, 2], [3, 4]], dtype=np.float64)
        scaler = StandardScaler()
        scaler.fit(X)
        scaler.dispose()
        with pytest.raises(DisposedError):
            scaler.transform(X)

    def test_column_mismatch(self):
        X = np.array([[1, 2], [3, 4]], dtype=np.float64)
        scaler = StandardScaler()
        scaler.fit(X)
        with pytest.raises(ValidationError, match='columns'):
            scaler.transform(np.zeros((1, 3)))

    def test_get_set_params(self):
        scaler = StandardScaler({'key': 'value'})
        assert scaler.get_params() == {'key': 'value'}
        scaler.set_params({'key': 'new'})
        assert scaler.get_params() == {'key': 'new'}


class TestMinMaxScaler:
    def test_fit_transform_scales_to_unit_range(self):
        X = np.array([[1, 10], [2, 20], [3, 30], [4, 40]], dtype=np.float64)
        scaler = MinMaxScaler()
        result = scaler.fit_transform(X)
        assert result.shape == X.shape
        assert np.allclose(result.min(axis=0), 0)
        assert np.allclose(result.max(axis=0), 1)

    def test_fit_then_transform(self):
        X = np.array([[0, 0], [5, 10], [10, 20]], dtype=np.float64)
        scaler = MinMaxScaler()
        scaler.fit(X)
        result = scaler.transform(X)
        expected = np.array([[0, 0], [0.5, 0.5], [1, 1]])
        assert np.allclose(result, expected)

    def test_save_load_roundtrip(self):
        X = np.array([[1, 10], [2, 20], [3, 30]], dtype=np.float64)
        scaler = MinMaxScaler()
        scaler.fit(X)
        result_orig = scaler.transform(X)

        bundle = scaler.save()
        manifest, toc, blobs = decode_bundle(bundle)
        loaded = MinMaxScaler._from_bundle(manifest, toc, blobs)
        result_loaded = loaded.transform(X)
        assert np.allclose(result_orig, result_loaded)

    def test_registry_load(self):
        X = np.array([[1, 2], [3, 4]], dtype=np.float64)
        scaler = MinMaxScaler()
        scaler.fit(X)
        bundle = scaler.save()

        loaded = registry_load(bundle)
        assert isinstance(loaded, MinMaxScaler)
        assert loaded.is_fitted

    def test_constant_column(self):
        X = np.array([[5, 1], [5, 2], [5, 3]], dtype=np.float64)
        scaler = MinMaxScaler()
        result = scaler.fit_transform(X)
        assert np.all(result[:, 0] == 0)

    def test_unseen_data_outside_range(self):
        X_train = np.array([[0], [10]], dtype=np.float64)
        scaler = MinMaxScaler()
        scaler.fit(X_train)
        X_test = np.array([[-5], [15]], dtype=np.float64)
        result = scaler.transform(X_test)
        assert result[0, 0] < 0
        assert result[1, 0] > 1

    def test_not_fitted_error(self):
        scaler = MinMaxScaler()
        with pytest.raises(NotFittedError):
            scaler.transform(np.zeros((1, 2)))

    def test_disposed_error(self):
        X = np.array([[1, 2], [3, 4]], dtype=np.float64)
        scaler = MinMaxScaler()
        scaler.fit(X)
        scaler.dispose()
        with pytest.raises(DisposedError):
            scaler.transform(X)

File: packages/core/test/test-scalers.js
const { describe, it } = require('node:test')
const assert = require('node:assert/strict')
const { StandardScaler, MinMaxScaler } = require('../src/scalers.js')
const { decodeBundle } = require('../src/bundle.js')
const { NotFittedError, DisposedError } = require('../src/errors.js')

function assertClose(a, b, tol = 1e-10) {
  assert.ok(Math.abs(a - b) < tol, `${a} not close to ${b} (tol=${tol})`)
}

describe('StandardScaler', () => {
  const X = {
    rows: 4,
    cols: 2,
    data: new Float64Array([
      1, 10,
      2, 20,
      3, 30,
      4, 40,
    ])
  }

  it('fit + transform produces zero-mean unit-variance columns', () => {
    const scaler = new StandardScaler()
    scaler.fit(X)
    const result = scaler.transform(X)

    assert.equal(result.rows, 4)
    assert.equal(result.cols, 2)

    // Check mean is ~0 for each column
    for (let c = 0; c < 2; c++) {
      let sum = 0
      for (let r = 0; r < 4; r++) sum += result.data[r * 2 + c]
      assertClose(sum / 4, 0, 1e-10)
    }

    // Check std is ~1 for each column
    for (let c = 0; c < 2; c++) {
      let mean = 0
      for (let r = 0; r < 4; r++) mean += result.data[r * 2 + c]
      mean /= 4
      let m2 = 0
      for (let r = 0; r < 4; r++) {
        const d = result.data[r * 2 + c] - mean
        m2 += d * d
      }
      const std = Math.sqrt(m2 / 3) // ddof=1
      assertClose(std, 1, 1e-10)
    }
  })

  it('fitTransform matches fit + transform', () => {
    const s1 = new StandardScaler()
    s1.fit(X)
    const r1 = s1.transform(X)

    const s2 = new StandardScaler()
    const r2 = s2.fitTransform(X)

    assert.deepEqual(r1.data, r2.data)
  })

  it('save + load round-trips', () => {
    const scaler = new StandardScaler()
    scaler.fit(X)
    const origResult = scaler.transform(X)

    const bytes = scaler.save()
    const { manifest, toc, blobs } = decodeBundle(bytes)
    assert.equal(manifest.typeId, 'wlearn.preprocess.standard_scaler@1')

    const loaded = StandardScaler._fromBundle(manifest, toc, blobs)
    const loadedResult = loaded.transform(X)

    assert.deepEqual(origResult.data, loadedResult.data)
  })

  it('handles number[][] input', () => {
    const scaler = new StandardScaler()
    scaler.fit([[1, 10], [2, 20], [3, 30], [4, 40]])
    const result = scaler.transform([[1, 10], [2, 20]])
    assert.equal(result.rows, 2)
    assert.equal(result.cols, 2)
  })

  it('handles constant column (std=0)', () => {
    const scaler = new StandardScaler()
    scaler.fit({ rows: 3, cols: 1, data: new Float64Array([5, 5, 5]) })
    const result = scaler.transform({ rows: 2, cols: 1, data: new Float64Array([5, 10]) })
    // Constant column -> output 0
    assert.equal(result.data[0], 0)
    assert.equal(result.data[1], 0)
  })

  it('throws NotFittedError before fit', () => {
    const scaler = new StandardScaler()
    assert.throws(() => scaler.transform(X), NotFittedError)
  })

  it('throws DisposedError after dispose', () => {
    const scaler = new StandardScaler()
    scaler.fit(X)
    scaler.dispose()
    assert.throws(() => scaler.transform(X), DisposedError)
    assert.equal(scaler.isFitted, false)
  })

  it('rejects column mismatch', () => {
    const scaler = new StandardScaler()
    scaler.fit(X)
    assert.throws(() => scaler.transform({ rows: 1, cols: 3, data: new Float64Array(3) }))
  })

  it('getParams and setParams work', () => {
    const scaler = new StandardScaler({ withMean: true })
    assert.deepEqual(scaler.getParams(), { withMean: true })
    scaler.setParams({ withMean: false })
    assert.deepEqual(scaler.getParams(), { withMean: false })
  })
})

describe('MinMaxScaler', () => {
  const X = {
    rows: 4,
    cols: 2,
    data: new Float64Array([
      1, 10,
      2, 20,
      3, 30,
      4, 40,
    ])
  }

  it('fit + transform scales to [0, 1]', () => {
    const scaler = new MinMaxScaler()
    scaler.fit(X)
    const result = scaler.transform(X)

    assert.equal(result.rows, 4)
    assert.equal(result.cols, 2)

    // First col: 1,2,3,4 -> 0, 1/3, 2/3, 1
    assertClose(result.data[0], 0)
    assertClose(result.data[2], 1 / 3)
    assertClose(result.data[4], 2 / 3)
    assertClose(result.data[6], 1)

    // Second col: 10,20,30,40 -> 0, 1/3, 2/3, 1
    assertClose(result.data[1], 0)
    assertClose(result.data[3], 1 / 3)
    assertClose(result.data[5], 2 / 3)
    assertClose(result.data[7], 1)
  })

  it('fitTransform matches fit + transform', () => {
    const s1 = new MinMaxScaler()
    s1.fit(X)
    const r1 = s1.transform(X)

    const s2 = new MinMaxScaler()
    const r2 = s2.fitTransform(X)

    assert.deepEqual(r1.data, r2.data)
  })

  it('save + load round-trips', () => {
    const scaler = new MinMaxScaler()
    scaler.fit(X)
    const origResult = scaler.transform(X)

    const bytes = scaler.save()
    const { manifest, toc, blobs } = decodeBundle(bytes)
    assert.equal(manifest.typeId, 'wlearn.preprocess.minmax_scaler@1')

    const loaded = MinMaxScaler._fromBundle(manifest, toc, blobs)
    const loadedResult = loaded.transform(X)

    assert.deepEqual(origResult.data, loadedResult.data)
  })

  it('handles constant column (range=0)', () => {
    const scaler = new MinMaxScaler()
    scaler.fit({ rows: 3, cols: 1, data: new Float64Array([5, 5, 5]) })
    const result = scaler.transform({ rows: 2, cols: 1, data: new Float64Array([5, 10]) })
    assert.equal(result.data[0], 0)
    assert.equal(result.data[1], 0)
  })

  it('can scale outside [0,1] for unseen data', () => {
    const scaler = new MinMaxScaler()
    scaler.fit({ rows: 2, cols: 1, data: new Float64Array([0, 10]) })
    const result = scaler.transform({ rows: 1, cols: 1, data: new Float64Array([20]) })
    assertClose(result.data[0], 2) // (20 - 0) / 10 = 2
  })

  it('throws NotFittedError before fit', () => {
    assert.throws(() => new MinMaxScaler().transform(X), NotFittedError)
  })

  it('throws DisposedError after dispose', () => {
    const scaler = new MinMaxScaler()
    scaler.fit(X)
    scaler.dispose()
    assert.throws(() => scaler.transform(X), DisposedError)
  })
})
